File: crates/psychevo-gateway/tests/fixtures/fake_acp_reconcile.py
#!/usr/bin/env python3
import json
import os
import sys

log_path = sys.argv[1]
state_path = sys.argv[2]


def load_state():
    try:
        with open(state_path, "r", encoding="utf-8") as state_file:
            return json.load(state_file)
    except (FileNotFoundError, json.JSONDecodeError):
        return {"promptCount": 0, "messages": []}


def save_state(state):
    with open(state_path, "w", encoding="utf-8") as state_file:
        json.dump(state, state_file)


def record(method, **fields):
    with open(log_path, "a", encoding="utf-8") as log_file:
        log_file.write(json.dumps({"method": method, **fields}) + "\n")


def send(value):
    print(json.dumps(value), flush=True)


def update(session_id, message_id, text):
    send(
        {
            "jsonrpc": "2.0",
            "method": "session/update",
            "params": {
                "sessionId": session_id,
                "update": {
                    "sessionUpdate": "agent_message_chunk",
                    "messageId": message_id,
                    "content": {"type": "text", "text": text},
                },
            },
        }
    )


for line in sys.stdin:
    if not line.strip():
        continue
    message = json.loads(line)
    method = message.get("method")
    mid = message.get("id")
    params = message.get("params") or {}
    if method == "initialize":
        record(method)
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "result": {
                    "protocolVersion": 1,
                    "agentCapabilities": {"loadSession": True},
                },
            }
        )
    elif method == "session/new":
        record(method)
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "result": {"sessionId": "native-reconcile"},
            }
        )
    elif method == "session/load":
        state = load_state()
        record(method, sessionId=params.get("sessionId"))
        send(
            {
                "jsonrpc": "2.0",
                "method": "session/update",
                "params": {
                    "sessionId": params.get("sessionId"),
                    "update": {
                        "sessionUpdate": "tool_call",
                        "toolCallId": "replayed-tool-only",
                        "title": "Replay tool-only fact",
                        "kind": "execute",
                        "status": "completed",
                    },
                },
            }
        )
        send(
            {
                "jsonrpc": "2.0",
                "method": "session/update",
                "params": {
                    "sessionId": params.get("sessionId"),
                    "update": {
                        "sessionUpdate": "plan",
                        "entries": [
                            {
                                "content": "Replay replacement plan",
                                "priority": "high",
                                "status": "completed",
                            }
                        ],
                    },
                },
            }
        )
        for replay in state["messages"]:
            update(params.get("sessionId"), replay["id"], replay["text"])
        send({"jsonrpc": "2.0", "id": mid, "result": {}})
    elif method == "session/prompt":
        state = load_state()
        state["promptCount"] += 1
        turn = state["promptCount"]
        prompt = "\n".join(
            block.get("text") or ""
            for block in params.get("prompt") or []
            if block.get("type") == "text"
        )
        answer = "reconciled answer " + str(turn)
        replay = {"id": "assistant-" + str(turn), "text": answer}
        state["messages"].append(replay)
        save_state(state)
        record(method, turn=turn, prompt=prompt)
        if turn == 1:
            os._exit(17)
        update(params.get("sessionId"), replay["id"], answer)
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "result": {"stopReason": "end_turn"},
            }
        )
    else:
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "error": {"code": -32601, "message": "method not found"},
            }
        )

File: crates/psychevo-gateway/tests/fixtures/fake_acp_reconcile.js
#!/usr/bin/env node
const fs = require("node:fs");
const readline = require("node:readline");
const logPath = process.argv[2];
const statePath = process.argv[3];
function loadState() { try { return JSON.parse(fs.readFileSync(statePath, "utf8")); } catch { return { promptCount: 0, messages: [] }; } }
function saveState(state) { fs.writeFileSync(statePath, JSON.stringify(state), "utf8"); }
function record(method, fields = {}) { fs.appendFileSync(logPath, `${JSON.stringify({ method, ...fields })}\n`, "utf8"); }
function send(value) { process.stdout.write(`${JSON.stringify(value)}\n`); }
function update(sessionId, messageId, text) { send({ jsonrpc: "2.0", method: "session/update", params: { sessionId, update: { sessionUpdate: "agent_message_chunk", messageId, content: { type: "text", text } } } }); }
function handle(message) {
  const { method, id } = message; const params = message.params || {};
  if (method === "initialize") { record(method); send({ jsonrpc: "2.0", id, result: { protocolVersion: 1, agentCapabilities: { loadSession: true } } }); }
  else if (method === "session/new") { record(method); send({ jsonrpc: "2.0", id, result: { sessionId: "native-reconcile" } }); }
  else if (method === "session/load") {
    const state = loadState(); record(method, { sessionId: params.sessionId });
    send({ jsonrpc: "2.0", method: "session/update", params: { sessionId: params.sessionId, update: { sessionUpdate: "tool_call", toolCallId: "replayed-tool-only", title: "Replay tool-only fact", kind: "execute", status: "completed" } } });
    send({ jsonrpc: "2.0", method: "session/update", params: { sessionId: params.sessionId, update: { sessionUpdate: "plan", entries: [{ content: "Replay replacement plan", priority: "high", status: "completed" }] } } });
    for (const replay of state.messages) update(params.sessionId, replay.id, replay.text);
    send({ jsonrpc: "2.0", id, result: {} });
  } else if (method === "session/prompt") {
    const state = loadState(); state.promptCount += 1; const turn = state.promptCount;
    const prompt = (params.prompt || []).filter((block) => block.type === "text").map((block) => block.text || "").join("\n");
    const answer = `reconciled answer ${turn}`; const replay = { id: `assistant-${turn}`, text: answer };
    state.messages.push(replay); saveState(state); record(method, { turn, prompt });
    if (turn === 1) process.exit(17);
    update(params.sessionId, replay.id, answer); send({ jsonrpc: "2.0", id, result: { stopReason: "end_turn" } });
  } else send({ jsonrpc: "2.0", id, error: { code: -32601, message: "method not found" } });
}
readline.createInterface({ input: process.stdin, crlfDelay: Infinity }).on("line", (line) => { if (line.trim()) handle(JSON.parse(line)); });
